<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="phyloraster">
<title>Introduction to the phyloraster package and its functionalities • phyloraster</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to the phyloraster package and its functionalities">
<meta property="og:description" content="phyloraster">
<meta property="og:image" content="https://gabferreira.github.io/phyloraster/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">phyloraster</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/phyloraster-vignette.html">Introduction to the phyloraster package and its functionalities</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Introduction to the phyloraster package and its functionalities</h1>
                        <h4 data-toc-skip class="author">Gabriela
Alves-Ferreira, Flávio Mota and Neander Heming</h4>
            
            <h4 data-toc-skip class="date">2023-11-30</h4>
      
      
      <div class="d-none name"><code>phyloraster-vignette.Rmd</code></div>
    </div>

    
    
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#install">Installation</a></li>
<li><a href="#data">Data processing</a></li>
<li><a href="#analysis">Analysis</a></li>
<li><a href="#null">Null Models</a></li>
<li><a href="#processing">Post Processing</a></li>
<li><a href="#references">References</a></li>
</ul>
<div class="section level2">
<h2 id="intro">Introduction<a class="anchor" aria-label="anchor" href="#intro"></a>
</h2>
<p><a href="https://github.com/gabferreira/phyloraster" class="external-link"><code>phyloraster</code></a>
is an R package to calculate measures of endemism and evolutionary
diversity using rasters of presence-absence as input, allowing to join
the results derived from species distribution models (SDMs) with
phylogenetic information. A lot of packages such as
<code>phyloregion</code> (Daru et al., 2020), <code>picante</code>
(Kembel et al., 2010) and <code>pez</code> (Pearse et al. 2015) can be
used to calculate patterns of endemism and evolutionary diversity.
However, most of these packages use matrices that can be computationally
expensive if the user is working with global or local data with high
resolution. <code>phyloraster</code> brings an alternative to these
packages by providing functions that calculate diversity and endemism
metrics for each raster cell, reducing the amount of RAM required for
data processing. The functions are focused on the steps of
pre-processing, processing and post-processing of macroecological and
phylogenetic data.</p>
<p>The pre-processing step offers basic functions for preparing the data
before running the analyses. The processing step brings together
functions to calculate Faith’s phylogenetic diversity, phylogenetic
endemism, weighted endemism and evolutionary distinctiveness. This step
also provides functions to calculate standardized effect size for each
metric through different methods of spatial and phylogenetic
randomization, aiming to control for richness effects. The user can
applying these methods of randomization to test hypotheses about the
community structure when richness patterns are correlated with endemism
and evolutionary patterns (Kembel et al. 2010). The post processing
stage includes functions to calculate the delta of metrics between
different times (e.g. present and future). We have shown that the
package has a slightly longer computation time than comparable packages,
but takes up a considerably smaller portion of RAM memory, which will
allow users to work with high-resolution datasets from local to global
scales. This enhances the application of the package by enabling users
to work with large datasets on computers with less RAM available. In
this vignette, we demonstrate the use of the functions of the
<code>phyloraster</code> package in detail.</p>
</div>
<div class="section level2">
<h2 id="install">Installation<a class="anchor" aria-label="anchor" href="#install"></a>
</h2>
<p>The CRAN version of the package can be installed using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"phyloraster"</span><span class="op">)</span></span></code></pre></div>
<p>The development version of <code>phyloraster</code> can be downloaded
from <a href="https://github.com/gabferreira/phyloraster" class="external-link">GitHub</a>
using the following code:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"gabferreira/phyloraster"</span><span class="op">)</span></span></code></pre></div>
<p>If you have any questions, let us know through the topic <a href="https://github.com/gabferreira/phyloraster/issues" class="external-link">“Issues”</a>.</p>
<p><code>phyloraster</code> uses some R packages as dependencies such as
<code>terra</code> (version &gt;= 1.6) (Hijmans, 2022), <code>ape</code>
(version &gt;= 5.6) (Paradis &amp; Schliep, 2019), and
<code>phylobase</code> (version &gt;= 0.810) (Hackathon et al. 2020).
Once installed, packages can be loaded into R using
<code><a href="https://rdrr.io/r/base/library.html" class="external-link">library()</a></code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://gabferreira.github.io/phyloraster/">phyloraster</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/" class="external-link">terra</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/fmichonneau/phylobase" class="external-link">phylobase</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data">Data processing<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>In the step of pre processing, we offer support to manipulate
matrices, shapefiles, rasters, and phylogenetic trees. In the processing
step, we provide functions to calculate Faith’s phylogenetic diversity
(Faith, 1992), phylogenetic endemism (Rosauer et al., 2009),
evolutionary distinctiveness (Isaac et al., 2007), and weighted endemism
(Williams et al., 1994).</p>
<ul>
<li><strong>Dataset</strong></li>
</ul>
<p>The package contains one dataset that allows visualizing the
structure expected to matrices, rasters, shapefiles and phylogenetic
trees and can be accessed using the function
<code><a href="../reference/load.data.rosauer.html">load.data.rosauer()</a></code>. This dataset contains a
<code>data.frame</code> with presence records for 33 Australian tree
frogs with coordinates for each site (Rosauer 2017) and a phylogenetic
tree for these species (Rosauer 2017). This raw dataset can be accessed
<a href="https://github.com/DanRosauer/phylospatial/tree/master/PhyloEndemism_in_R/Tree%20Frog%20Data" class="external-link">here</a>.
The function also provide a binary raster of presence absence and a
shapefile with the range of 27 species following IUCN spatial data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load.data.rosauer.html">load.data.rosauer</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">presab</span><span class="op">)</span></span>
<span><span class="co">#&gt;   Longitude Latitude Litoria_revelata Litoria_rothii Litoria_longirostris</span></span>
<span><span class="co">#&gt; 1  144.0657  -14.894                0              1                    0</span></span>
<span><span class="co">#&gt; 2  144.0657  -15.194                0              1                    0</span></span>
<span><span class="co">#&gt; 3  144.0657  -16.694                0              1                    0</span></span>
<span><span class="co">#&gt; 4  144.0657  -15.494                0              1                    0</span></span>
<span><span class="co">#&gt; 5  144.0657  -16.294                0              1                    0</span></span>
<span><span class="co">#&gt; 6  144.0657  -16.594                0              1                    0</span></span>
<span><span class="co">#&gt;   Litoria_dorsalis Litoria_rubella Litoria_inermis Litoria_pallida</span></span>
<span><span class="co">#&gt; 1                0               1               1               1</span></span>
<span><span class="co">#&gt; 2                0               1               1               1</span></span>
<span><span class="co">#&gt; 3                0               1               1               1</span></span>
<span><span class="co">#&gt; 4                0               1               1               1</span></span>
<span><span class="co">#&gt; 5                0               1               1               1</span></span>
<span><span class="co">#&gt; 6                0               1               1               1</span></span>
<span><span class="co">#&gt;   Litoria_latopalmata Litoria_tornieri Litoria_nasuta Litoria_nigrofrenata</span></span>
<span><span class="co">#&gt; 1                   0                0              1                    0</span></span>
<span><span class="co">#&gt; 2                   0                0              1                    0</span></span>
<span><span class="co">#&gt; 3                   0                0              1                    0</span></span>
<span><span class="co">#&gt; 4                   0                0              1                    0</span></span>
<span><span class="co">#&gt; 5                   0                0              1                    0</span></span>
<span><span class="co">#&gt; 6                   0                0              1                    0</span></span>
<span><span class="co">#&gt;   Litoria_bicolor Litoria_fallax Litoria_genimaculata Litoria_wilcoxii</span></span>
<span><span class="co">#&gt; 1               1              0                    0                0</span></span>
<span><span class="co">#&gt; 2               0              0                    0                0</span></span>
<span><span class="co">#&gt; 3               0              0                    0                0</span></span>
<span><span class="co">#&gt; 4               0              0                    0                0</span></span>
<span><span class="co">#&gt; 5               0              0                    0                0</span></span>
<span><span class="co">#&gt; 6               0              0                    0                0</span></span>
<span><span class="co">#&gt;   Litoria_jungguy Litoria_andiirrmalin Litoria_caerulea Litoria_chloris</span></span>
<span><span class="co">#&gt; 1               0                    0                1               0</span></span>
<span><span class="co">#&gt; 2               0                    0                1               0</span></span>
<span><span class="co">#&gt; 3               0                    0                1               0</span></span>
<span><span class="co">#&gt; 4               0                    0                1               0</span></span>
<span><span class="co">#&gt; 5               0                    0                1               0</span></span>
<span><span class="co">#&gt; 6               0                    0                1               0</span></span>
<span><span class="co">#&gt;   Litoria_xanthomera Litoria_gracilenta Cyclorana_brevipes</span></span>
<span><span class="co">#&gt; 1                  0                  0                  1</span></span>
<span><span class="co">#&gt; 2                  0                  0                  1</span></span>
<span><span class="co">#&gt; 3                  0                  0                  1</span></span>
<span><span class="co">#&gt; 4                  0                  1                  1</span></span>
<span><span class="co">#&gt; 5                  0                  0                  1</span></span>
<span><span class="co">#&gt; 6                  0                  0                  1</span></span>
<span><span class="co">#&gt;   Cyclorana_novaehollandiae Cyclorana_cultripes Cyclorana_manya</span></span>
<span><span class="co">#&gt; 1                         1                   0               1</span></span>
<span><span class="co">#&gt; 2                         1                   0               0</span></span>
<span><span class="co">#&gt; 3                         1                   0               1</span></span>
<span><span class="co">#&gt; 4                         1                   0               1</span></span>
<span><span class="co">#&gt; 5                         1                   0               1</span></span>
<span><span class="co">#&gt; 6                         1                   0               1</span></span>
<span><span class="co">#&gt;   Litoria_alboguttata Cyclorana_longipes Nyctimystes_dayi Litoria_nannotis</span></span>
<span><span class="co">#&gt; 1                   1                  0                0                0</span></span>
<span><span class="co">#&gt; 2                   1                  0                0                0</span></span>
<span><span class="co">#&gt; 3                   1                  0                0                0</span></span>
<span><span class="co">#&gt; 4                   1                  0                0                0</span></span>
<span><span class="co">#&gt; 5                   1                  0                0                0</span></span>
<span><span class="co">#&gt; 6                   1                  0                0                0</span></span>
<span><span class="co">#&gt;   Litoria_lorica Litoria_rheocola Litoria_nyakalensis Litoria_infrafrenata</span></span>
<span><span class="co">#&gt; 1              0                0                   0                    0</span></span>
<span><span class="co">#&gt; 2              0                0                   0                    0</span></span>
<span><span class="co">#&gt; 3              0                0                   0                    0</span></span>
<span><span class="co">#&gt; 4              0                0                   0                    0</span></span>
<span><span class="co">#&gt; 5              0                0                   0                    0</span></span>
<span><span class="co">#&gt; 6              0                0                   0                    0</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span><span class="op">$</span><span class="va">tree</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Phylogenetic tree with 33 tips and 26 internal nodes.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Tip labels:</span></span>
<span><span class="co">#&gt;   Litoria_revelata, Litoria_rothii, Litoria_longirostris, Litoria_dorsalis, Litoria_rubella, Litoria_nigrofrenata, ...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Rooted; includes branch lengths.</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">tree</span>, cex <span class="op">=</span> <span class="fl">0.65</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/unnamed-chunk-6-1.png" width="480" style="display: block; margin: auto;"></p>
<ul>
<li>
<strong>Function</strong> <code>df2rast</code>
</li>
</ul>
<p>The function <code>df2rast</code> converts traditional communities
matrices (i.e. species in columns and sites in rows, with coordinates in
the two first columns) into binary distribution rasters (presence and
absence).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load.data.rosauer.html">load.data.rosauer</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">r</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/df2rast.html">df2rast</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">presab</span>, </span>
<span>             CRS <span class="op">=</span> <span class="st">"+proj=longlat +datum=WGS84 +ellps=WGS84 +towgs84=0,0,0"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "SpatRaster"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "terra"</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">r</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/unnamed-chunk-8-1.png" width="480" style="display: block; margin: auto;"></p>
<ul>
<li>
<strong>Function</strong> <code>shp2rast</code>
</li>
</ul>
<p>The <code>shp2rast</code> function transform a shapefile to a raster
stack with the same extent. This function allows to work, for example,
with the shapes of species distribution provided by the <a href="https://www.iucnredlist.org/resources/spatial-data-download" class="external-link">International
Union for the Conservation of Nature’s Spatial Database</a>. We provide
a set of shapefiles for 27 species of Australian tree frogs. You can
visualize this data through the following code:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">shp</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/vect.html" class="external-link">vect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"shps_iucn_spps_rosauer.shp"</span>, </span>
<span>                               package <span class="op">=</span> <span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">rainbow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">shp</span><span class="op">$</span><span class="va">BINOMIAL</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                  alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="va">position</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/terra/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">shp</span><span class="op">$</span><span class="va">BINOMIAL</span>,</span>
<span>                  <span class="fu"><a href="https://rdrr.io/pkg/terra/man/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">shp</span><span class="op">$</span><span class="va">BINOMIAL</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="va">colors</span><span class="op">[</span><span class="va">position</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">shp</span>, col <span class="op">=</span> <span class="va">colors</span>, lty <span class="op">=</span> <span class="fl">0</span>,</span>
<span>     main <span class="op">=</span> <span class="st">"Spatial polygons"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">maps</span><span class="op">)</span></span>
<span><span class="fu">maps</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/maps/man/map.html" class="external-link">map</a></span><span class="op">(</span>add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/unnamed-chunk-10-1.png" width="480" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">r2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/shp2rast.html">shp2rast</a></span><span class="op">(</span><span class="va">shp</span>, sps.col <span class="op">=</span> <span class="st">"BINOMIAL"</span>, ymask <span class="op">=</span> <span class="cn">FALSE</span>, background <span class="op">=</span> <span class="fl">0</span>, </span>
<span>               resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="va">r2</span></span>
<span><span class="co">#&gt; class       : SpatRaster </span></span>
<span><span class="co">#&gt; dimensions  : 77, 126, 9  (nrow, ncol, nlyr)</span></span>
<span><span class="co">#&gt; resolution  : 0.5, 0.5  (x, y)</span></span>
<span><span class="co">#&gt; extent      : 114.099, 177.099, -39.17965, -0.6796477  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#&gt; coord. ref. : lon/lat WGS 84 </span></span>
<span><span class="co">#&gt; source(s)   : memory</span></span>
<span><span class="co">#&gt; names       : Litor~ensis, Litor~ttata, Litor~ensis, Litor~malin, Litor~giana, Litor~kiana, ... </span></span>
<span><span class="co">#&gt; min values  :           0,           0,           0,           0,           0,           0, ... </span></span>
<span><span class="co">#&gt; max values  :           1,           1,           1,           1,           1,           1, ...</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">r2</span><span class="op">[[</span><span class="fl">9</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>You can also masking the shapefile using another shapefile, as
follows:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/" class="external-link">terra</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">shp</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/vect.html" class="external-link">vect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"shps_iucn_spps_rosauer.shp"</span>,</span>
<span>                              package<span class="op">=</span><span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create a polygon to use as mask with an extent</span></span>
<span><span class="va">e</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/ext.html" class="external-link">ext</a></span><span class="op">(</span><span class="fl">113</span>, <span class="fl">123</span>, <span class="op">-</span><span class="fl">43.64</span>, <span class="op">-</span><span class="fl">33.90</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/as.polygons.html" class="external-link">as.polygons</a></span><span class="op">(</span><span class="va">e</span>, crs<span class="op">=</span><span class="st">""</span><span class="op">)</span></span>
<span><span class="co"># cut by the total extension of the polygons</span></span>
<span><span class="va">coun.crop</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/crop.html" class="external-link">crop</a></span><span class="op">(</span><span class="va">p</span>, </span>
<span>                         <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/ext.html" class="external-link">ext</a></span><span class="op">(</span><span class="va">shp</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">coun.rast</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/rasterize.html" class="external-link">rasterize</a></span><span class="op">(</span><span class="va">coun.crop</span>,</span>
<span><span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/ext.html" class="external-link">ext</a></span><span class="op">(</span><span class="va">shp</span><span class="op">)</span>, resolution <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rasterizing with the mask of the polygon</span></span>
<span><span class="va">shp.t</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/shp2rast.html">shp2rast</a></span><span class="op">(</span><span class="va">shp</span>, y <span class="op">=</span> <span class="va">coun.rast</span>, sps.col <span class="op">=</span> <span class="st">"BINOMIAL"</span>, ymask <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">shp.t</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="st">"green"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<ul>
<li>
<strong>Function</strong> <code>phylo.pres</code>
</li>
</ul>
<p>To calculate evolutionary measurements it is extremely important that
the raster with species distributions and the tree have the species
names in the same order. We can verify this premise using a simple
logical test like this.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load.data.rosauer.html">load.data.rosauer</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/print-methods.html" class="external-link">names</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">raster</span><span class="op">)</span> <span class="op">==</span> <span class="va">data</span><span class="op">$</span><span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span></span>
<span><span class="co">#&gt;  [1]  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE  TRUE</span></span>
<span><span class="co">#&gt; [13]  TRUE  TRUE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE  TRUE  TRUE FALSE</span></span>
<span><span class="co">#&gt; [25] FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE</span></span></code></pre></div>
<p>The function <code>phylo.pres</code> reorder the raster stack
according to phylogenetic tree order, extract a subtree containing only
species present in the raster stack and get the branch length for each
species.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ras</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rast.presab.tif"</span>, </span>
<span>                               package <span class="op">=</span> <span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"tree.nex"</span>, </span>
<span>                                   package <span class="op">=</span> <span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dataprep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phylo.pres.html">phylo.pres</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ras</span>, tree <span class="op">=</span> <span class="va">tree</span><span class="op">)</span></span></code></pre></div>
<p>Now, the raster stack and the tip label of the tree are in the same
order!</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/print-methods.html" class="external-link">names</a></span><span class="op">(</span><span class="va">dataprep</span><span class="op">$</span><span class="va">x</span><span class="op">)</span> <span class="op">==</span> <span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span></span>
<span><span class="co">#&gt;  [1] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">#&gt; [16] TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE TRUE</span></span>
<span><span class="co">#&gt; [31] TRUE TRUE TRUE</span></span></code></pre></div>
<p>The user also has the option to compute branch length and descendant
number using the full supplied tree or the tree subsetted by the species
present in the raster. Notice the implications of using the full or the
subsetted tree. Consider, for instance, a scenario where a clade
comprises four species (A, B, C and D - Figure 1a), and the particular
area of study involves two of these species (A and B, in blue- Figure
1). Furthermore, let’s assume that species A and B share a branch,
denoted as D (in red- Figure 1). Using the full phylogenetic tree will
estimate the whole length of branches for these two species, including
the branch shared between them (D), that connects them with the ancestor
shared with the species absent from that specific region (Figure 1b). On
the other hand, when using the subsetted tree (Figure 1c), branch D will
be disregarded and only the terminal branches will be used to calculate
branch length, so that the calculated branch lengths of the species A
and B will be shorter (Figure 1c).</p>
<p><img src="figs/tree.jpg" style="display: block; margin: auto;">
Figure 1. Phylogenetic tree for tree frogs denoting the implications of
using the full or the subsetted tree in the phylo.pres function. Figure
a) demonstrates the full phylogenetic tree for some tree frog species.
In figure b), we have the full tree considering species A and B that are
present in the region (blue), as well as the common ancestor between the
two species (red). In figure c) we only have the species present in the
region.</p>
</div>
<div class="section level2">
<h2 id="analysis">Analysis<a class="anchor" aria-label="anchor" href="#analysis"></a>
</h2>
<p>Great!! Now, we are already able to calculate the measures of species
richness, endemism and evolutionary diversity.</p>
<ul>
<li><strong>Species richness</strong></li>
</ul>
<p>Our package allows you to calculate species richness using the
<code>rast.sr</code> function.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ras</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rast.presab.tif"</span>, </span>
<span>                               package <span class="op">=</span> <span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rast.sr.html">rast.sr</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ras</span><span class="op">)</span></span>
<span><span class="va">sr</span></span>
<span><span class="co">#&gt; class       : SpatRaster </span></span>
<span><span class="co">#&gt; dimensions  : 90, 68, 1  (nrow, ncol, nlyr)</span></span>
<span><span class="co">#&gt; resolution  : 0.1, 0.1  (x, y)</span></span>
<span><span class="co">#&gt; extent      : 144.0157, 150.8157, -23.044, -14.044  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#&gt; coord. ref. : lon/lat WGS 84 (EPSG:4326) </span></span>
<span><span class="co">#&gt; source(s)   : memory</span></span>
<span><span class="co">#&gt; name        : SR </span></span>
<span><span class="co">#&gt; min value   :  2 </span></span>
<span><span class="co">#&gt; max value   : 29</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">sr</span>, main <span class="op">=</span> <span class="st">"Species richness"</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/plot-1.png" width="672" style="display: block; margin: auto;"></p>
<ul>
<li><strong>Endemism measurements</strong></li>
</ul>
<p>The <code>phyloraster</code> package implements functions for
calculating spatial patterns of endemism based on the weighted endemism
method (WE; Williams et al. 1994, Crisp et al. 2001) through the
function <code>rast.we</code>. The function returns a raster with the
values of weighted endemism by each pixel. Endemism values range from 0
to 1.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ras</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rast.presab.tif"</span>, </span>
<span>                               package <span class="op">=</span> <span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">wer</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rast.we.html">rast.we</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ras</span><span class="op">)</span></span>
<span><span class="va">wer</span></span>
<span><span class="co">#&gt; class       : SpatRaster </span></span>
<span><span class="co">#&gt; dimensions  : 90, 68, 1  (nrow, ncol, nlyr)</span></span>
<span><span class="co">#&gt; resolution  : 0.1, 0.1  (x, y)</span></span>
<span><span class="co">#&gt; extent      : 144.0157, 150.8157, -23.044, -14.044  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#&gt; coord. ref. : lon/lat WGS 84 (EPSG:4326) </span></span>
<span><span class="co">#&gt; source(s)   : memory</span></span>
<span><span class="co">#&gt; name        :           WE </span></span>
<span><span class="co">#&gt; min value   : 0.0007231441 </span></span>
<span><span class="co">#&gt; max value   : 0.2989257007</span></span></code></pre></div>
<p>By using the R plot function from <code>terra</code> package it is
possible to visualize the regions where species with restricted range
are distributed.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wer</span><span class="op">$</span><span class="va">WE</span></span>
<span><span class="co">#&gt; class       : SpatRaster </span></span>
<span><span class="co">#&gt; dimensions  : 90, 68, 1  (nrow, ncol, nlyr)</span></span>
<span><span class="co">#&gt; resolution  : 0.1, 0.1  (x, y)</span></span>
<span><span class="co">#&gt; extent      : 144.0157, 150.8157, -23.044, -14.044  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#&gt; coord. ref. : lon/lat WGS 84 (EPSG:4326) </span></span>
<span><span class="co">#&gt; source(s)   : memory</span></span>
<span><span class="co">#&gt; name        :           WE </span></span>
<span><span class="co">#&gt; min value   : 0.0007231441 </span></span>
<span><span class="co">#&gt; max value   : 0.2989257007</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">wer</span><span class="op">$</span><span class="va">WE</span>, main <span class="op">=</span><span class="st">"Weigthed Endemism"</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/unnamed-chunk-19-1.png" width="672" style="display: block; margin: auto;"></p>
<ul>
<li><strong>Evolutionary measurements</strong></li>
</ul>
<p>The first evolutionary measure is Faith’s phylogenetic diversity (PD,
Faith 1994), which is calculated as the sum of the branch length for all
species occurring in a given region (Faith 1994).</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ras</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rast.presab.tif"</span>, </span>
<span>                               package <span class="op">=</span> <span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"tree.nex"</span>, </span>
<span>                                   package <span class="op">=</span> <span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dataprep</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phylo.pres.html">phylo.pres</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ras</span>, tree <span class="op">=</span> <span class="va">tree</span>, pruning <span class="op">=</span> <span class="st">"tree"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pdr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rast.pd.html">rast.pd</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dataprep</span><span class="op">$</span><span class="va">x</span>, edge.path <span class="op">=</span> <span class="va">dataprep</span><span class="op">$</span><span class="va">edge.path</span>, </span>
<span>               branch.length <span class="op">=</span> <span class="va">dataprep</span><span class="op">$</span><span class="va">branch.length</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pdr</span><span class="op">$</span><span class="va">PD</span>, main <span class="op">=</span> <span class="st">"Phylogenetic diversity"</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/pdr-plot-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The second measure is phylogenetic endemism (PE, Rosauer et
al. 2009), which calculates the degree to which PD are restricted to a
specific region (Rosauer et al. 2009). The function <code>rast.pe</code>
returns a raster layer containing PE the region of interest.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ras</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rast.presab.tif"</span>, </span>
<span>                               package <span class="op">=</span> <span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"tree.nex"</span>, </span>
<span>                                   package <span class="op">=</span> <span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">per</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rast.pe.html">rast.pe</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">dataprep</span><span class="op">$</span><span class="va">x</span>, <span class="va">tree</span><span class="op">)</span></span>
<span><span class="va">per</span></span>
<span><span class="co">#&gt; class       : SpatRaster </span></span>
<span><span class="co">#&gt; dimensions  : 90, 68, 1  (nrow, ncol, nlyr)</span></span>
<span><span class="co">#&gt; resolution  : 0.1, 0.1  (x, y)</span></span>
<span><span class="co">#&gt; extent      : 144.0157, 150.8157, -23.044, -14.044  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">#&gt; coord. ref. : lon/lat WGS 84 (EPSG:4326) </span></span>
<span><span class="co">#&gt; source(s)   : memory</span></span>
<span><span class="co">#&gt; name        :           PE </span></span>
<span><span class="co">#&gt; min value   : 0.0002224064 </span></span>
<span><span class="co">#&gt; max value   : 0.1756404428</span></span></code></pre></div>
<p>The result can be visualized using the R <code>plot</code> function
from the <code>terra</code> package.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">per</span><span class="op">$</span><span class="va">PE</span>, main <span class="op">=</span> <span class="st">"Phylogenetic Endemism"</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/per-plot-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="null">Null models<a class="anchor" aria-label="anchor" href="#null"></a>
</h2>
<p>Null models are a widely used method to control for richness effects
in diversity measures (Gotelli and Ulrich, 2012). The standardized
effect size (SES) measure, also known as z-score or z-value, is used to
calculate null models from randomization tests (Gotelli and McCabe
2002). <code>phyloraster</code> implements three methods to calculate
SES using spatial and phylogenetic randomization: rast.pe.ses,
rast.pd.ses, and rast.we.ses.</p>
<ul>
<li><strong>Spatial and phylogenetic randomization</strong></li>
</ul>
<p>The randomization procedure for the calculation of SES is done
internally in the functions <code><a href="../reference/rast.we.ses.html">rast.we.ses()</a></code>,
<code><a href="../reference/rast.pd.ses.html">rast.pd.ses()</a></code>, <code><a href="../reference/rast.ed.ses.html">rast.ed.ses()</a></code>,
<code><a href="../reference/rast.pe.ses.html">rast.pe.ses()</a></code>, and <code><a href="../reference/geo.phylo.ses.html">geo.phylo.ses()</a></code> through the
package <code>SESraster</code> (Heming et al., 2023).
<code>SESraster</code> currently implements six algorithms to randomize
binary species distribution with several levels of constraints: SIM1,
SIM2, SIM3, SIM5, SIM6, and SIM9 (sensu Gotelli, 2000). The methods
implemented in the <code>SESraster</code> are based on how species
(originally rows) and sites (originally columns) are treated
(i.e. fixed, equiprobable, or proportional sums) (Gotelli, 2000). The
randomization algorithms currently available in <code>SESraster</code>
are: SIM1 (species occurrence equiprobable and site richness
equiprobable), SIM2 (species occurrence fixed and site richness
equiprobable), SIM3 (species occurrence equiprobable and site richness
fixed), SIM5 (species occurrence proportional and site richness fixed),
SIM6 (species occurrence proportional and site richness fixed) and SIM9
(species occurrence fixed and site richness fixed, similar to the
preserved model of Laffan &amp; Crisp, 2003). In addition,
<code>SESraster</code> (consequently <code>phyloraster</code>) supports
user’s custom randomization algorithms for SES calculation, as long as
the function returns objects of class SpatRaster. This allows complete
flexibility for using any algorithm not yet implemented by the
package.</p>
<p>As default, the <code>phyloraster</code> uses the function
<code><a href="https://hemingnm.github.io/SESraster/reference/bootspat_str.html" class="external-link">bootspat_str()</a></code> from the <code>SESraster</code> package to
conduct the randomizations, but the user is free to choose any of the
other methods mentioned above through the spat_alg argument in the
<code>*.ses()</code> functions of the <code>phyloraster</code> package.
The function <code><a href="https://hemingnm.github.io/SESraster/reference/bootspat_str.html" class="external-link">bootspat_str()</a></code> is equivalent to the SIM5
(proportional-fixed) method of Gotelli (2000), which partially relaxes
the spatial structure of species distributions but keeps the spatial
structure of the observed richness pattern across cells.</p>
<p>Now that we have presented the randomization methods, we can start to
build the null models.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://CRAN.R-project.org/package=SESraster" class="external-link">SESraster</a></span><span class="op">)</span></span>
<span><span class="va">ras</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rast.presab.tif"</span>, </span>
<span>                               package <span class="op">=</span> <span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"tree.nex"</span>, </span>
<span>                                   package <span class="op">=</span> <span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phylo.pres.html">phylo.pres</a></span><span class="op">(</span><span class="va">ras</span>, <span class="va">tree</span>, pruning <span class="op">=</span> <span class="st">"tree"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">t</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rast.pd.ses.html">rast.pd.ses</a></span><span class="op">(</span><span class="va">data</span><span class="op">$</span><span class="va">x</span>, edge.path <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">edge.path</span>, </span>
<span>                 branch.length <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">branch.length</span>, aleats <span class="op">=</span> <span class="fl">10</span>, </span>
<span>                 random <span class="op">=</span> <span class="st">"spat"</span><span class="op">)</span></span></code></pre></div>
<p>Plotting the results</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">t</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/pds-plot-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="processing">Post processing<a class="anchor" aria-label="anchor" href="#processing"></a>
</h2>
<ul>
<li>
<strong>Function</strong> <code>delta.grid</code>
</li>
</ul>
<p>The package also brings the function <code>delta.grid</code> that
allows you to calculate the difference between spatialized diversity
metrics between two different times. This function would allow assessing
how species richness varies between different times, which could be
useful in a climate change scenario. For example, imagine that we
currently have 33 tree frog species of the subfamily Pelodryadinae
occurring in Australia. In the map below we can visualize the spatial
pattern of species richness.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the data</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rast.presab.tif"</span>, </span>
<span>                             package<span class="op">=</span><span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># richness</span></span>
<span><span class="va">riq.pres</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rast.sr.html">rast.sr</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">riq.pres</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/sr-plot-1.png" width="700"></p>
<p>Now imagine that as climate change progresses, 16 relatively more
vulnerable species are heavily affected and become locally extinct.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the data</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/rast.html" class="external-link">rast</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"rast.presab.tif"</span>, </span>
<span>                             package<span class="op">=</span><span class="st">"phyloraster"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># richness future</span></span>
<span><span class="va">riq.fut</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rast.sr.html">rast.sr</a></span><span class="op">(</span><span class="va">x</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="co"># imagine we lost some species in the future</span></span>
<span><span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">riq.fut</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/srf-plot-1.png" width="700"></p>
<p>The <code>delta.grid</code> function allows you to visualize the
variation in these richness patterns spatially. See an example
below.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/delta.grid.html">delta.grid</a></span><span class="op">(</span><span class="va">riq.pres</span>, <span class="va">riq.fut</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">dg</span><span class="op">)</span></span></code></pre></div>
<p><img src="phyloraster-vignette_files/figure-html/dg-1.png" width="700"></p>
<p>On the map we can see that the greatest loss occurs in the eastern
region of the map, losing up to 16 species. The <code>delta.grid</code>
function can be used for any of the other metrics that are available in
the <code>phyloraster</code> package.</p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<p>Crisp, M., Laffan, S., Linder, H. and Monro, A. (2001). Endemism in
the Australian flora. Journal of Biogeography, 28, 183–198.</p>
<p>Daru, B. H., Karunarathne, P., and Schliep, K. (2020). phyloregion: R
package for biogeographical regionalization and macroecology. Methods in
Ecology and Evolution, 11(11), 1483–1491. <a href="https://doi.org/10.1111/2041-210X.13478" class="external-link uri">https://doi.org/10.1111/2041-210X.13478</a></p>
<p>Faith, D. P. (1992). Conservation evaluation and phylogenetic
diversity. Biological conservation, 61(1), 1-10.</p>
<p>Gotelli, N. J., and McCabe, D. J. (2002). Species co-occurrence: A
meta-analysis of J. M. Diamond’s assembly rules model. Ecology, 83(8),
2091–2096. <a href="https://doi.org/10.1890/0012-9658(2002)083%5B2091:SCOAMA%5D2.0.CO;2" class="external-link uri">https://doi.org/10.1890/0012-9658(2002)083[2091:SCOAMA]2.0.CO;2</a></p>
<p>Gotelli, N. J., and Ulrich, W. (2012). Statistical challenges in null
model analysis. Oikos, 121(2), 171–180. <a href="https://doi.org/10.1111/j.1600-0706.2011.20301.x" class="external-link uri">https://doi.org/10.1111/j.1600-0706.2011.20301.x</a></p>
<p>Hackathon, R. (2020). phylobase: Base Package for Phylogenetic
Structures and Comparative Data (0.8. 10). <a href="https://CRAN.R-project.org/package=phylobase" class="external-link uri">https://CRAN.R-project.org/package=phylobase</a></p>
<p>Hjimans, R. J. (2022). Terra, Spatial Data Analysis (1.6.7). <a href="https://CRAN.R-project.org/package=terra" class="external-link uri">https://CRAN.R-project.org/package=terra</a></p>
<p>Isaac, N. J., Turvey, S. T., Collen, B., Waterman, C. and Baillie, J.
E. (2007). Mammals on the EDGE: conservation priorities based on threat
and phylogeny. PLoS ONE 2, e296.</p>
<p>Kembel, S. W., Cowan, P. D., Helmus, M. R., Cornwell, W. K., Morlon,
H., Ackerly, D. D., Blomberg, S. P., and Webb, C. O. (2010). Picante: R
tools for integrating phylogenies and ecology. Bioinformatics, 26(11),
1463–1464. <a href="https://doi.org/10.1093/bioinformatics/btq166" class="external-link uri">https://doi.org/10.1093/bioinformatics/btq166</a></p>
<p>Paradis, E., and Schliep, K. (2019). ape 5.0: An environment for
modern phylogenetics and evolutionary analyses in R. Bioinformatics, 35,
526–528.</p>
<p>Pearse, W. D., Cadotte, M. W., Cavender-Bares, J., Ives, A. R.,
Tucker, C. M., Walker, S. C., &amp; Helmus, M. R. (2015). pez:
Phylogenetics for the environmental sciences. Bioinformatics, 31(17),
2888–2890. <a href="https://doi.org/10.1093/bioinformatics/btv277" class="external-link uri">https://doi.org/10.1093/bioinformatics/btv277</a></p>
<p>Rosauer, D. A. N., Laffan, S. W., Crisp, M. D., Donnellan, S. C. and
Cook, L. G. (2009). Phylogenetic endemism: a new approach for
identifying geographical concentrations of evolutionary history.
Molecular ecology, 18(19), 4061-4072.</p>
<p>Williams, P.H., Humphries, C.J., Forey, P.L., Humphries, C.J. and
VaneWright, R.I. (1994). Biodiversity, taxonomic relatedness, and
endemism in conservation. In: Systematics and Conservation Evaluation
(eds Forey PL, Humphries CJ, Vane-Wright RI), p. 438. Oxford University
Press, Oxford.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Gabriela Alves-Ferreira, Flávio M. M. Mota, Neander Marcel Heming.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
